File: LinFu.DesignByContract2.Examples/VBContractAssertions/Assertions/Invariants/BalanceIsAlwaysNonNegativeAttribute.vb
Imports LibraryInterfaces
Imports LinFu.DesignByContract2.Core

<AttributeUsage(AttributeTargets.Class Or AttributeTargets.Interface)> _
Public Class BalanceIsAlwaysNonNegativeAttribute
    Inherits Attribute
    Implements IInvariant

    Public Function AppliesTo(ByVal target As Object, ByVal info As LinFu.DynamicProxy.InvocationInfo) As Boolean Implements LinFu.DesignByContract2.Core.IContractCheck.AppliesTo
        If target Is Nothing Then Return False
        If Not (TypeOf target Is IBankAccount) Then Return False
        Return True
    End Function

    Public Sub [Catch](ByVal ex As System.Exception) Implements LinFu.DesignByContract2.Core.IContractCheck.Catch

    End Sub

    Public Function Check(ByVal target As Object, ByVal info As LinFu.DynamicProxy.InvocationInfo, ByVal callState As LinFu.DesignByContract2.Core.InvariantState) As Boolean Implements LinFu.DesignByContract2.Core.IInvariant.Check
        Dim account As IBankAccount = CType(target, IBankAccount)
        Return account.Balance >= 0
    End Function

    Public Sub ShowError(ByVal output As System.IO.TextWriter, ByVal target As Object, ByVal info As LinFu.DynamicProxy.InvocationInfo, ByVal callState As LinFu.DesignByContract2.Core.InvariantState) Implements LinFu.DesignByContract2.Core.IInvariant.ShowError
        output.WriteLine("You cannot have a negative balance!")
    End Sub
End Class

File: LinFu.DesignByContract2.Examples/ContractAssertions/Assertions/Invariants/BalanceIsAlwaysNonNegativeAttribute.cs
using System;
using System.Collections.Generic;
using System.Text;
using LibraryInterfaces;
using LinFu.DesignByContract2.Core;
using LinFu.DynamicProxy;

namespace CSContractAssertions
{
    [AttributeUsage(AttributeTargets.Interface | AttributeTargets.Class)]
    public class BalanceIsAlwaysNonNegativeAttribute : Attribute, IInvariant
    {
        #region IInvariant Members

        public bool Check(object target, InvocationInfo info, InvariantState callState)
        {
            IBankAccount account = (IBankAccount) target;
            return account.Balance >= 0;
        }

        public void ShowError(System.IO.TextWriter output, object target, InvocationInfo info, InvariantState callState)
        {
            output.WriteLine("You cannot have a negative balance!");
        }

        #endregion

        #region IContractCheck Members

        public bool AppliesTo(object target, InvocationInfo info)
        {
            return target != null && target is IBankAccount;
        }

        public void Catch(Exception ex)
        {
            // Do nothing
        }

        #endregion
    }
}
